[Feature][Discussion] Add support for legends to calendars as another view

Raistlfiren opened the pull request:
Hi, 

This is in its infancy, but I would like to propose a new feature for the calendar. Simply put, I just wanted to add a calendar legend onto the screen. This would be more applicable for users displaying a monthly calendar with multiple different calendars attached to it. Currently, all it does is the following - 

- Gets the calendar configurations of all the calendars from `config.js`.
- A new reference would be under the views array for `legend`
```javascript 
{
  name: "Legend",
  mode: "legend",
  position: "top_left",
  calendars: ["Holidays", "Garmin", "EMG"]
}
```
- Displays a new object or div on the screen of the calendars where positioned in the `config.js`.
- Markup looks like this (If there is a class under the calendar config then it is added to legend-slot):
```html
<div class="module fake_module MMM-CalendarExt2-Legend CX2 shown">
   <div class="module-content">
      <div class="view legend">
         <div class="legend-slot" data-calendar-name="Holidays">
            <div class="iconify eventIcon svg-loading" data-icon="calendar-check"></div>
            <div class="eventTitle">Holidays</div>
         </div>
         <div class="legend-slot" data-calendar-name="Garmin">
            <div class="eventTitle">Garmin</div>
         </div>
         <div class="legend-slot" data-calendar-name="EMG">
            <div class="eventTitle">EMG</div>
         </div>
      </div>
   </div>
</div>
```
- End result looks like this:
<img width="486" alt="Screen Shot 2022-03-31 at 8 55 16 AM" src="https://user-images.githubusercontent.com/234712/161072262-d0c15ea2-3003-4992-bbca-25f14d7cfb07.png">

Raistlfiren commented:
@klaernie - Thank you for the response. My end result was a bad example and probably didn't provide good enough context to understand the purpose behind things. A better example would be from a previous thread on the MagicMirror forums (not my screenshot but what I am trying to replicate natively with the code - the bottom right legend of calendars) -  https://forum.magicmirror.builders/topic/11001/config-for-a-simple-clean-montly-calendar/5

Mostly, it is just to easily identify different calendars on the monthly calendar view. Although, a possible secondary use case would be to easily filter the calendar upon what the user "touched" on the screen too. The plus side of including it with the CalendarExt2 in another view is to get the user's calendar configuration and base the legend on that setup. The image showcased above is using a static image that isn't easily updated.

klaernie commented:
Okay, that explains the use case for this view type fairly clearly.
Does the style of the legend item change with the same selectors as the main events do?

Else I'd mainly would like to see that the new view is documented like the others are.

Raistlfiren commented:
@klaernie The HTML markup is as shown - 

```html
<div class="module fake_module MMM-CalendarExt2-Legend CX2 shown">
   <div class="module-content">
      <div class="view legend">
         <div class="legend-slot" data-calendar-name="Holidays">
            <div class="iconify eventIcon svg-loading" data-icon="calendar-check"></div>
            <div class="eventTitle">Holidays</div>
         </div>
         <div class="legend-slot" data-calendar-name="Garmin">
            <div class="eventTitle">Garmin</div>
         </div>
         <div class="legend-slot" data-calendar-name="EMG">
            <div class="eventTitle">EMG</div>
         </div>
      </div>
   </div>
</div>
```

The main modification to the HTML is changed the module class from `MMM-CalendarExt2` to `MMM-CalendarExt2-Legend`. I added the class `legend` to the parent class of the legend items, and added `legend-slot` to each item. If there is a `className` associated with the calendar then it is appended to each item. For example - `div class="legend-slot <className>" ...>`. Should I modify the HTML to have the same class as the other views or do you want me to modify the HTML to be more unique like changing `eventIcon` to `legendIcon` and `eventTitle` to `legendTitle`?

I will definitely start working on documentation for the view tonight. Thank you @klaernie .

klaernie commented:
> do you want me to modify the HTML to be more unique like changing `eventIcon` to `legendIcon` and `eventTitle` to `legendTitle`

Precisely the opposite - avoid uniqueness to the legend as far as possible.
My rationale is, that I expect the legend to require minimal configuration - as far as we can sensibly expect.
So going by "what we have in docs/" (aka just my personal config at the moment (which I just for this PR refreshed), feel free to add more examples) I'd expect users to start their per-calendar styling at `.CX2 .event[data-calendar-name="…"]{…}`. And with that I'd expect that the legend only needs to be added, and will automatically adjust to the calendars currently in view and display every legend entries like events of that calendar would be. I hope this makes sense to your.

There's also another gotcha you're probably not aware of: the deduplication code will generate calendar-names that are combinations of the other calendar names (enabled with https://github.com/klaernie/docker-MagicMirror/blob/main/doc/examples/k8s/klaernie/xl-config.yaml#L78-L80, typical calendar names at https://github.com/klaernie/docker-MagicMirror/blob/main/doc/examples/k8s/klaernie/xl-config.yaml#L121-L137).

Raistlfiren commented:
@klaernie - I updated my code to be more structured like you requested. I got rid of the `MMM-CalendarExt2-Legend` and replaced it with `MMM-CalendarExt2`. I also added the `event` class to each legend item, and added a header for the legend similar to the other views. There is also documentation for the view too. 

In regards to de-duplication, I am unsure how it is handled.  Currently, I am just looking to see if the legend view has the calendars in the configuration and pull that information. You can see my logic under `CALEXT2_Scene.js`. I mostly just loop through the config calendars and see if the names match the calendars in the legend view. I am not sure what the de-duplication view looks like.

Thanks!

klaernie commented:
> There is also documentation for the view too.

Cool! But one thing maybe: the typical end user will not care for the HTML that is output, only for what they can tweak or set, and how it will look and behave. So maybe this would be a good place to explain to the user, what the legend accomplishes.

> Currently, I am just looking to see if the legend view has the calendars in the configuration and pull that information

Ah, I think I now understand: if you have two scenes (like in my example the "private calendars" and "work calendars" scenes) you also need to have two legend views and include them in the corresponding scene.
This makes it a little less magic than I was hoping for, but it would be fine to leave it at that - after all it exactly matches the way the other views work across scenes.

Please also take a look at the linter output again:
The CSS [has a few issues](https://github.com/MMM-CalendarExt2/MMM-CalendarExt2/runs/5801640176?check_suite_focus=true#step:5:100) (you should be able to run `npm run lint` locally to discover these).
[It also complains](https://github.com/MMM-CalendarExt2/MMM-CalendarExt2/runs/5801640176?check_suite_focus=true#step:5:240) about the singular space at the start of line 5 in `legend.md`

> I am not sure what the de-duplication view looks like.

Deduplication isn't a view, but a process during the calendar ingest, merging events found in multiple calendars, so there is just one event instead of the original events (think a family lunch that's in every family members calendar, so there's no point in showing three entries of this singular event, just because everyone will partake). If the list of calendars would be loaded from the views that are currently in view (that's what I was hoping on being magically done above) the deduplicated events would have calendar names that are combinations, which could be confusing at least.
But as the logic is right now this won't be a problem, since the user exactly specifies which calendars are listed in the legend.

Raistlfiren commented:
Thank you @klaernie for all of the useful information . I updated the documentation to be more precise in my wording, and I also fixed the issues for linter. (I hope) :)

klaernie commented on docs/Configuration/Views/legend.md:
```diff
@@ -0,0 +1,52 @@
+<!-- markdownlint-disable-file MD041 -->
+
+# View:legend
+The purpose of this view is to show a legend of all calendars specified in the configuration to display. The legend is typically used on a large calendar view to denote different calendars being used. For example, you might have a Google calendar for work and another for home. The legend will allow you or a guest to quickly differentiate events on your calendar from home or work. This makes more  ...
```
please reformat to about 80 char width

klaernie commented on docs/Configuration/Views/legend.md:
```diff
@@ -0,0 +1,52 @@
+# View:legend
+The purpose of this view is to show a legend of all calendars specified in the configuration to display. The legend is typically used on a large calendar view to denote different calendars being used. For example, you might have a Google calendar for work and another for home. The legend will allow you or a guest to quickly differentiate events on your calendar from home or work. This makes more  ...
+
+### Configuration Reference
```
please make this a `##` header

klaernie commented on docs/Configuration/Views/legend.md:
````diff
@@ -0,0 +1,52 @@
+### Legend with Calendar
+![Full Legend Calendar Example](legend-full-example.png)
+
+The outputted HTML is very simplistic and looks like the following:
+```html
+<div class="module fake_module MMM-CalendarExt2 CX2 shown">
+  <header class="module-header">Legend</header>
+  <div class="module-content">
+    <div class="view legend">
+      <div class="legendSlot event garmin" data-calendar-name="Garmin">
+        <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" width="1em" height="1em" style="vertical-align: -0.125em;-ms-transform: rotate(360deg); -webkit-transform: rotate(360deg); transform: rotate(360deg);" preserveAspectRatio="xMidYMid meet" viewBox="0 0 512 512" class="iconify eventIcon" data-icon="cil:running">
+          <path fill="currentColor" d="m353.415 200l-30.981-57.855l-60.717-20.239l-.14.432L167.21 149.3A32.133 32.133 0 0 0 144 180.068V264h32v-83.931l73.6-21.028l-32.512 99.633l-.155-.056l-29.464 82.5a16.088 16.088 0 0 1-20.127 9.8l-66.282-22.097l-10.12 30.358l66.282 22.093a48 48 0 0 0 60.378-29.391l24.232-67.849l17.173 5.6l48.3 48.3A15.9 15.9 0 0 1 312 349.255V456h32V349.255a47.694 47.694 0 0 0 ...
+        </svg>
+        <div class="eventTitle">Garmin</div>
+      </div>
+      <div class="legendSlot event work" data-calendar-name="Work">
+        <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" width="1em" height="1em" style="vertical-align: -0.125em;-ms-transform: rotate(360deg); -webkit-transform: rotate(360deg); transform: rotate(360deg);" preserveAspectRatio="xMidYMid meet" viewBox="0 0 24 24" class="iconify eventIcon" data-icon="gg:work-alt">
+          <path fill="currentColor" fill-rule="evenodd" d="M17 7a3 3 0 0 0-3-3h-4a3 3 0 0 0-3 3H6a3 3 0 0 0-3 3v8a3 3 0 0 0 3 3h12a3 3 0 0 0 3-3v-8a3 3 0 0 0-3-3h-1Zm-3-1h-4a1 1 0 0 0-1 1h6a1 1 0 0 0-1-1ZM6 9h12a1 1 0 0 1 1 1v8a1 1 0 0 1-1 1H6a1 1 0 0 1-1-1v-8a1 1 0 0 1 1-1Z" clip-rule="evenodd"></path>
+        </svg>
+        <div class="eventTitle">Work</div>
+      </div>
+      <div class="legendSlot event home" data-calendar-name="Home">
+        <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" width="1em" height="1em" style="vertical-align: -0.125em;-ms-transform: rotate(360deg); -webkit-transform: rotate(360deg); transform: rotate(360deg);" preserveAspectRatio="xMidYMid meet" viewBox="0 0 1024 1024" class="iconify eventIcon" data-icon="ant-design:home-outlined">
+          <path fill="currentColor" d="M946.5 505L560.1 118.8l-25.9-25.9a31.5 31.5 0 0 0-44.4 0L77.5 505a63.9 63.9 0 0 0-18.8 46c.4 35.2 29.7 63.3 64.9 63.3h42.5V940h691.8V614.3h43.4c17.1 0 33.2-6.7 45.3-18.8a63.6 63.6 0 0 0 18.7-45.3c0-17-6.7-33.1-18.8-45.2zM568 868H456V664h112v204zm217.9-325.7V868H632V640c0-22.1-17.9-40-40-40H432c-22.1 0-40 17.9-40 40v228H238.1V542.3h-96l370-369.7l23.1 23.1L882 ...
+        </svg>
+        <div class="eventTitle">Home</div>
+      </div>
+    </div>
+  </div>
+</div>
+```
````
Please remove this HTML entirely. The users will not care how the legend does it's magic, and it will be outdated fairly soon. This then also fixes the last linter error.

Raistlfiren replied:
I suppose the HTML isn't necessary, but it is a good reference to see what CSS classes need to be updated with loading everything for the legend or calendar.

klaernie replied:
For that the Chrome Devtools with an inspector are always the better option, since it allows to play with the styling at the same time.

klaernie reviewed (changes requested):
please make a few changes to the docs, and once the linter passes I'll merge it.

Raistlfiren commented:
Thanks @klaernie. I will try to update things this week or the weekend.

Raistlfiren commented:
@klaernie - These issues should now be resolved. Please review and I will correct anything further.
